src/buffer.rs

```diff
@@ -0,0 +1,301 @@
+use crate::channel::mpsc::Sender as PrioritySender;
+use futures::channel::mpsc::Sender;
+use std::{
+    collections::VecDeque,
+    task::{Context, Poll},
+};
+
+/// If the buffer capacity is greater than u8 max, shrink it
+const BUF_SHRINK_THRESHOLD: usize = u8::max_value() as usize;
+
+pub enum SendResult {
+    Ok,
+    Pending,
+    Disconnect,
+}
+
+pub struct PriorityBuffer<T> {
+    sender: PrioritySender<T>,
+    high_buffer: VecDeque<T>,
+    normal_buffer: VecDeque<T>,
+}
+
+impl<T> PriorityBuffer<T> {
+    pub fn new(sender: PrioritySender<T>) -> Self {
+        PriorityBuffer {
+            sender,
+            high_buffer: VecDeque::default(),
+            normal_buffer: VecDeque::default(),
+        }
+    }
+
+    pub fn push_high(&mut self, item: T) {
+        self.high_buffer.push_back(item)
+    }
+
+    pub fn push_normal(&mut self, item: T) {
+        self.normal_buffer.push_back(item)
+    }
+
+    pub fn len(&self) -> usize {
+        self.high_buffer.len() + self.normal_buffer.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.high_buffer.is_empty() && self.normal_buffer.is_empty()
+    }
+
+    fn shrink_to_fit(&mut self) {
+        if self.high_buffer.capacity() > self.high_buffer.len() + BUF_SHRINK_THRESHOLD {
+            self.high_buffer.shrink_to_fit();
+        }
+        if self.normal_buffer.capacity() > self.normal_buffer.len() + BUF_SHRINK_THRESHOLD {
+            self.normal_buffer.shrink_to_fit();
+        }
+    }
+
+    pub fn try_send(&mut self, cx: &mut Context) -> SendResult {
+        while let Some(event) = self.high_buffer.pop_front() {
+            match self.sender.poll_ready(cx) {
+                Poll::Ready(Ok(())) => {
+                    if let Err(e) = self.sender.try_quick_send(event) {
+                        if e.is_full() {
+                            self.high_buffer.push_front(e.into_inner());
+                            return SendResult::Pending;
+                        } else {
+                            return SendResult::Disconnect;
```

Comment:
Log the error?

Reply:
There is no log inside the structure definition, only when it is used, it is necessary to consider whether to add

Reply:
If it's OK to add a String in `SendResult::Disconnect`, we can add the reason in it.

Reply:
There is only one reason for disconnect, the sender end is closed/dropped